File: py-usfm-parser/src/usfm_grammar/markers_ext_reader.py
'''Inputs a filepath or file contents of the format
\\marker zmyp
\\category versepara
\\description An paragraph marker extension.

or 

\\marker zmyc
\\category char
\\description A character marker extension.
\\attribute x-myattr1

Reads this content and builds a dictionary/json object.
The value of the marker field becomes the key and 
each of the other lines in that block becomes a key-value pair in the value object.
'''
import re

type_map = {
    'para': 'para',
    'header': 'para',
    'title': 'para',
    'introduction': 'para',
    'section': 'para',
    'para': 'para',
    'versepara': 'para',
    'list': 'para',
    'otherpara': 'para',
    'note': 'note',
    'crossreference': 'note',
    'footnote': 'note',
    'char': 'char',
    'introchar': 'char',
    'listchar': 'char',
    'footnotechar': 'char',
    'crossreferencechar': 'char',
    'milestone': 'milestone',
}

replacement_map = {
    'para': 'customPara_',
    'char': 'customChar_',
    'note': 'customNote_',
    'milestone': 'customMS_',
}


line_pattern = re.compile(r'\\([\w\-]+)\s+(.*)')
class ExtensionReader:
    def __init__(self):
        self.lines = []
        self.extensions = {}

    def read_to_object(self,file_content=None, file_path=None):    
        if file_path and file_content is None:
            file_content = open(file_path, 'r', encoding='utf-8').read()
        self.lines = file_content.splitlines()

        current_marker = None
        for line in self.lines:
            line_match = re.match(line_pattern, line)
            if line_match is not None:
                key = line_match.group(1)
                value = line_match.group(2)
                if key == "marker":
                    if value.startswith('z'):
                        self.extensions[value] = {}
                        current_marker = value
                    else:
                        current_marker = None
                elif current_marker is not None:
                    self.extensions[current_marker][key] = value 
            else:
                pass
                # print(f"Line not conforming to pattern:{line}")
        return self.extensions

    def replace_custom_markers(self, usfm_string):
        modified_usfm = usfm_string
        for marker in self.extensions:
            marker_type = type_map[self.extensions[marker]['category']]
            replacement = replacement_map[marker_type]
            # Replace the marker with the replacement prefix followed by the original marker 
            # if the marker is enclosed by a backslash and a space, newline or *
            marker_pattern = re.compile(rf"\\{re.escape(marker)}(?=[^\w\-]|$)")

            modified_usfm = marker_pattern.sub(
                lambda m: f"\\{replacement}{marker}",
                modified_usfm,
            )
        return modified_usfm

File: node-usfm-parser/src/markersExtReader.js
const typeMap = {
  para: 'para',
  header: 'para',
  title: 'para',
  introduction: 'para',
  section: 'para',
  versepara: 'para',
  list: 'para',
  otherpara: 'para',
  note: 'note',
  crossreference: 'note',
  footnote: 'note',
  char: 'char',
  introchar: 'char',
  listchar: 'char',
  footnotechar: 'char',
  crossreferencechar: 'char',
  milestone: 'milestone',
};

const replacementMap = {
  para: 'customPara_',
  char: 'customChar_',
  note: 'customNote_',
  milestone: 'customMS_',
};

const linePattern = /^\\([\w-]+)\s+(.*)$/;

class ExtensionReader {
  constructor() {
    this.lines = [];
    this.extensions = {};
  }

  readToObject(fileContent = null, filePath = null) {
    if (fileContent === null && filePath !== null) {
      const fs = require('fs');
      fileContent = fs.readFileSync(filePath, 'utf8');
    }
    if (fileContent === null) {
      throw new TypeError('fileContent is required.');
    }
    this.lines = fileContent.split(/\r?\n/);
    this.extensions = {};
    let currentMarker = null;
    for (const line of this.lines) {
      const lineMatch = line.match(linePattern);
      if (lineMatch === null) {
        continue;
      }
      const [, key, value] = lineMatch;
      if (key === 'marker') {
        if (value.startsWith('z')) {
          this.extensions[value] = {};
          currentMarker = value;
        } else {
          currentMarker = null;
        }
      } else if (currentMarker !== null) {
        this.extensions[currentMarker][key] = value;
      }
    }
    return this.extensions;
  }

  replaceCustomMarkers(usfmString) {
    let modifiedUsfm = usfmString;
    for (const marker of Object.keys(this.extensions)) {
      const category = this.extensions[marker].category;
      const markerType = typeMap[category];
      if (markerType === undefined) {
        continue;
      }
      const replacement = replacementMap[markerType];
      const markerPattern = new RegExp(`\\\\${marker}(?=[^\\w-]|$)`, 'g');
      modifiedUsfm = modifiedUsfm.replace(
        markerPattern,
        `\\${replacement}${marker}`,
      );
    }
    return modifiedUsfm;
  }
}

module.exports = ExtensionReader;
module.exports.ExtensionReader = ExtensionReader;
module.exports.replacementMap = replacementMap;
module.exports.typeMap = typeMap;